src/osd/scrubber/scrub_reservations.cc

```diff
@@ -169,23 +204,24 @@ bool ReplicaReservations::send_next_reservation_or_complete()
 
   // send the next reservation request
   const auto peer = *m_next_to_request;
-  const auto epoch = m_pg->get_osdmap_epoch();
+  const auto ep = m_pg->get_osdmap_epoch();
+  m_last_request_sent_nonce++;
+
   auto m = make_message<MOSDScrubReserve>(
-      spg_t{m_pgid, peer.shard}, epoch, MOSDScrubReserve::REQUEST,
-      m_pg->pg_whoami);
-  m_pg->send_cluster_message(peer.osd, m, epoch, false);
+      spg_t{m_pgid, peer.shard}, ep, MOSDScrubReserve::REQUEST, m_pg->pg_whoami,
+      m_last_request_sent_nonce);
+  m_pg->send_cluster_message(peer.osd, m, ep, false);
   m_last_request_sent_at = ScrubClock::now();
   dout(10) << fmt::format(
-		  "reserving {} (the {} of {} replicas)", *m_next_to_request,
-		  active_requests_cnt() + 1, m_sorted_secondaries.size())
+		  "reserving {} (the {} of {} replicas) e:{} id:{}",
+		  *m_next_to_request, active_requests_cnt() + 1,
+		  m_sorted_secondaries.size(), ep, m_last_request_sent_nonce)
 	   << dendl;
   m_next_to_request++;
   return false;
 }
 
-void ReplicaReservations::verify_rejections_source(
-    OpRequestRef op,
-    pg_shard_t from)
+bool ReplicaReservations::handle_rejection(OpRequestRef op, pg_shard_t from)
```

Comment:
This doesn't "handle" the request, it's a predicate checking recency and origin.  Rename it to is_rejection_valid and modify the signature to take a reference to the message type rather than an OpRequestRef.

Reply:
Also, we have to apply the same check to a success, right?

Reply:
> This doesn't "handle" the request, it's a predicate checking recency and origin

It's more than that now: previously, the FSM called both log_failure_and_duration() and verify_rejections_source(). Now - handle_rejection() does both (as there was no reason to require a separate API to ReservingReplica). If you do not like the generic 'handle' - maybe check_and_log_rejection()?

Reply:
> signature to take a reference to the message type rather than an OpRequestRef.

done

Reply:
> Also, we have to apply the same check to a success, right?

we do ?

Reply:
Oh, I see.  ScrubMachine gets distinct ReplicaReject and ReplicaGrant messages, so we can't really check the nonce in a single place -- that's unfortunate but would require more refactoring than is really worthwhile here to correct.

handle_rejection does have a state change I missed -- how about handle_reserve_rejection to make it symmetric to handle_reserve_grant?